src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs

```diff
@@ -0,0 +1,136 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IClientMessageInspector"/> implementation which adds telemetry to outgoing requests.
+    /// </summary>
+    public class TelemetryClientMessageInspector : IClientMessageInspector
+    {
+        /// <inheritdoc/>
+        public object BeforeSendRequest(ref Message request, IClientChannel channel)
+        {
+            try
+            {
+                if (WcfInstrumentationActivitySource.Options == null || WcfInstrumentationActivitySource.Options.OutgoingRequestFilter?.Invoke(request) == false)
+                {
+                    WcfInstrumentationEventSource.Log.RequestIsFilteredOut(request.Headers.Action);
+                    return new State
+                    {
+                        SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                    };
+                }
+            }
+            catch (Exception ex)
+            {
+                WcfInstrumentationEventSource.Log.RequestFilterException(ex);
+                return new State
+                {
+                    SuppressionScope = this.SuppressDownstreamInstrumentation(),
+                };
+            }
+
+            Activity activity = WcfInstrumentationActivitySource.ActivitySource.StartActivity(
+                WcfInstrumentationActivitySource.OutgoingRequestActivityName,
+                ActivityKind.Client);
+
+            IDisposable suppressionScope = this.SuppressDownstreamInstrumentation();
+
+            if (activity != null)
+            {
+                string action = request.Headers.Action;
+                activity.DisplayName = action;
+
+                WcfInstrumentationActivitySource.Options.Propagator.Inject(
+                    new PropagationContext(activity.Context, Baggage.Current),
+                    request,
+                    WcfInstrumentationActivitySource.MessageHeaderValueSetter);
+
+                if (activity.IsAllDataRequested)
+                {
+                    int lastIndex = action.LastIndexOf('/');
+
+                    activity.SetTag(WcfInstrumentationConstants.RpcSystemTag, WcfInstrumentationConstants.WcfSystemValue);
+                    activity.SetTag(WcfInstrumentationConstants.RpcServiceTag, action.Substring(0, lastIndex));
+                    activity.SetTag(WcfInstrumentationConstants.RpcMethodTag, action.Substring(lastIndex + 1));
```

Comment:
This isn't always valid. This will work for any auto-generated action names, and it will work for most manually created names but there's no requirement for this, it's purely a convention. An alternative might be to look up the operation based on the action name then use the actual method name. The same with RpcServiceTag, you could look up the interface name. I suspect interface name and method signature (because of overloads) would be a better fit here. If so, let me know and I'll clone this repo and try work out what code you would need to be able to retrieve that info here. Basically you need to capture a this information from the runtime when adding the message inspector then add a dictionary to the message inspector for this to be looked up. Even then you need to handle wild cards as you can have a method which accepts any action.

Reply:
@mconnew If you don't mind helping out that would be awesome! I just added you as a collaborator on the fork. Not sure if you wanted to add on this PR or do one after it merges.

Reply:
I'll try and get to this on this PR but don't block on me as I can submit a new PR afterwards if needed.

Reply:
So it looks like there's some api's missing in our contract on .NET Core and can only do this for Task based methods. So you could either fetch them by reflection (which will be stable as long as you use public or nonpublic on the reflection call) or do a fallback behavior to use ClientOperation.Name when the operation isn't a Task based one. This has a problem of not differentiating overloads though. I coded up the fallback way. Here's the code that I put into the TelemetryEndpointBehavior.ApplyClientBehavior method. I didn't submit a PR as you might want to break this down into a struct instead of appending it all into a single string, but this is the mechanism you would use anyway.
```csharp
            var actionMapping = new Dictionary<string, string>();
            foreach (var operation in clientRuntime.ClientOperations)
            {
                var action = operation.Action;
                string methodName = operation.Name;
#if NETFRAMEWORK
                var method = operation.SyncMethod ?? operation.TaskMethod;
                if (method == null)
                {
                    // Protect against bad extensibility which can break operation
                    if (operation.BeginMethod != null && operation.EndMethod != null)
                    {
                        methodName = operation.BeginMethod.DeclaringType.FullName + ":" +
                                     operation.BeginMethod.ToString() + ";" + operation.EndMethod.ToString();
                    }
                    else
                    {
                        continue;
                    }
                }
                else
                {
                    methodName = method.DeclaringType.FullName + ":" + method.ToString();
                }
#else
                var method = operation.TaskMethod;
                if (method != null)
                {
                    methodName = method.DeclaringType.FullName + ":" + method.ToString();
                }
                else
                {
                    methodName = operation.Name;
                }
#endif // NETFRAMEWORK

                actionMapping[action] = methodName;
            }

            this.telemetryClientMessageInspector.ActionMapping = actionMapping.ToImmutableDictionary();
```

I added an ImmutableDictionary\<string,string\> called ActionMapping into the message inspector. Then in the message inspector you can do a lookup of the action. ClientRuntime.SyncMethod, BeginMethod, and EndMethod aren't in the contract but are in the implementation. If you open an issue in dotnet/wcf to request these be exposed (so you get notified when it happens), I can get these added to the next release. I ddn't know how comfortable you would be with a reflection based implementation which is why I didn't push changes to your fork.

Reply:
@mconnew Thanks for the code! I opened an issue on dotnet/wcf for you wrt to the API gap. I was playing with this a bit locally. I actually don't think we need to resolve down to the MethodInfo that is handling the operation. I implemented a version that uses the above concept to build a cache of all the contract name + operation name combos. The reason I think that works is it more closely matches what the other OpenTelemetry instrumentations are doing. For example, in http we put things like the route template on the spans not the action handler that actually received the request.